app/uk/gov/hmrc/exports/models/declaration/submissions/Action.scala

```diff
@@ -18,17 +18,70 @@ package uk.gov.hmrc.exports.models.declaration.submissions
 
 import play.api.libs.functional.syntax._
 import play.api.libs.json._
+import uk.gov.hmrc.exports.models.declaration.ExportsDeclaration
+import uk.gov.hmrc.exports.models.declaration.submissions.RequestType.RequestTypeFormat
 
 import java.time.{LocalDateTime, ZoneId, ZonedDateTime}
 
-case class Action(
-  id: String,
-  requestType: RequestType,
-  requestTimestamp: ZonedDateTime = ZonedDateTime.now(ZoneId.of("UTC")),
-  notifications: Option[Seq[NotificationSummary]] = None
-) {
+sealed trait Action {
```

Comment:
I do not understand the reason for turning the original Action case class into a trait and then created a lot of specialist implementation case classes for all the different RequestTypes. Apart from being able to hard code the `SubmissionAction` `val versionNo: Int = 1` there is no benefit, and consequently it leads to a lot of duplicate code (as all the other Action case class implementations are exactly the same as each other).

Reply:
I agreed with Tim. It seems more an academic exercise.

Reply:
May I suggest to have a look [at this PR](https://github.com/hmrc/customs-declare-exports/pull/421/files#diff-9906e4255892c4389b4066892891b3c1cf71f90ccb52c119ed0e6f7329e35ef5)?
 I'm also using for the migration an ad-hoc `reads` to add nested fields to the model, but the approach is different from what your are doing.

Reply:
It provides more than this.
Each action has different requirements according to the `requestType`, so on reading the `requestType` and assigning to an appropriate case class allows us to better define the Action. 
For example the `versionNo` of `SubmissionAction` will always be 1, furthermore an `ExternalAmendmentAction` will have an `decId: Option[String]` but all other Actions must have `decId: String`
It also helps when instantiating Actions, where a custom apply for CancellationAction can ensure that the `versionNo` will be equal to the `submission.latestVersionNo` whereas a custom apply for `AmendmentAction` will iterate on `submission.latestVersionNo`
By using traits, types, custom apply methods we can implement all those rules that appear complicated in the story
It can make things more readable too

Reply:
Yesterday I was tired and it was too late to continue with the review :woozy_face:
Sorry Darryl, but I don't see anything in the ticket requirements really convincing to justify this multiplication of case classes. According to your logic we should then have 10 different `ExportsDeclaration` case classes based on journey type and additional type.
I would agree with your approach if the structure of the classes would require different fields but that's not case here. The structure is unique for all different `requestType`.

Reply:
The requirements were for:
An action with a requestType == SubmissionRequest to have decId == submission.uuid and a versionNo == 1
An action with a requestType == CancellationRequest to have decId == submission.latestDecId and a submission.latestVersionNo
An action with a requestType == AmendmentRequest to have decId == declaration.id and a submission.latestVersionNo + 1
An action with a requestType == ExternalAmendmentRequest to have an optional decId and a submission.latestVersionNo + 1

Each action hinges on its requestType and each has different requirements
This approach allows us to to enforce all these rules from the ticket... whats the problem?